compiler-core/src/type_/tests.rs

```diff
@@ -1678,6 +1678,85 @@ fn unused_imported_module_warnings_test() {
     );
 }
 
+#[test]
+fn duplicate_qualified_imports_test() {
+    let mut warnings = vec![];
+    let ids = UniqueIdGenerator::new();
+    let mut modules = im::HashMap::new();
+
+    // DUPE: preludeinsertion
+    // TODO: Currently we do this here and also in the tests. It would be better
+    // to have one place where we create all this required state for use in each
+    // place.
+    let _ = modules.insert("gleam".to_string(), build_prelude(&ids));
+
+    let (mut ast, _) = crate::parse::parse_module("pub fn bar() { 1 }").expect("syntax error");
+    ast.name = vec!["gleam".to_string(), "foo".to_string()];
+    let module = infer_module(
+        Target::Erlang,
+        &ids,
+        ast,
+        Origin::Src,
+        "thepackage",
+        &modules,
+        &mut warnings,
+    )
+    .expect("should successfully infer");
+    let _ = modules.insert(
+        vec!["gleam".to_string(), "foo".to_string()].join("/"),
+        module.type_info,
+    );
+
+    let (mut ast, _) = crate::parse::parse_module("pub fn bar() { 2 }").expect("syntax error");
+    ast.name = vec!["gleam2".to_string(), "foo".to_string()];
+    let module = infer_module(
+        Target::Erlang,
+        &ids,
+        ast,
+        Origin::Src,
+        "thepackage",
+        &modules,
+        &mut warnings,
+    )
+    .expect("should successfully infer");
+    let _ = modules.insert(
+        vec!["gleam2".to_string(), "foo".to_string()].join("/"),
+        module.type_info,
+    );
+
+    let (mut ast, _) = crate::parse::parse_module(
+        "
+    import gleam/foo
+    import gleam2/foo
+
+    pub fn main() {
+        foo.bar()
+    }
+    ",
+    )
+    .expect("syntax error");
+    ast.name = vec!["my_module".to_string()];
+    let infer_result = infer_module(
+        Target::Erlang,
+        &ids,
+        ast,
+        Origin::Src,
+        "thepackage",
+        &modules,
+        &mut warnings,
+    );
+
+    assert!(infer_result.is_err());
+    assert_eq!(
+        Error::DuplicateImport {
+            location: SrcSpan { start: 33, end: 43 },
```

Comment:
Is there a way to test this "e2e"? So that we can check that the spans are correct and so on.

Reply:
You could create a module error test using the macro in the type tests modules.

https://github.com/gleam-lang/gleam/blob/2e27150eade0bf192753079059eb88bbd48ce29b/compiler-core/src/type_/tests/errors.rs#L1699-L1701